src/gui/src/drcWidget.cpp

```diff
@@ -0,0 +1,538 @@
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <QDebug>
+
+#include <QHeaderView>
+#include <QVBoxLayout>
+#include <QFileDialog>
+
+#include <map>
+#include <fstream>
+#include <regex>
+#include <sstream>
+#include <iomanip>
+
+#include "drcWidget.h"
+#include "utl/Logger.h"
+
+Q_DECLARE_METATYPE(gui::DRCViolation*);
+
+namespace gui {
+
+static QString getBBoxString(const odb::Rect& bbox, odb::dbBlock* block)
+{
+  double to_microns = block->getDbUnitsPerMicron();
+  std::stringstream ss;
+  ss << std::fixed << std::setprecision(5) << "(";
+  ss << bbox.xMin() / to_microns << ",";
+  ss << bbox.yMin() / to_microns << "), (";
+  ss << bbox.xMax() / to_microns << ",";
+  ss << bbox.yMax() / to_microns << ")";
+  return QString::fromStdString(ss.str());
+}
+
+///////
+
+DRCViolation::DRCViolation(const std::string& type,
+                           const std::vector<std::any>& srcs,
+                           const std::vector<DRCShape>& shapes,
+                           odb::dbTechLayer* layer,
+                           const std::string& comment) :
+                               type_(type),
+                               srcs_(srcs),
+                               shapes_(shapes),
+                               layer_(layer),
+                               comment_(comment)
+{
+  computeBBox();
+}
+
+DRCViolation::DRCViolation(const std::string& type,
+                           const std::vector<DRCShape>& shapes,
+                           const std::string& comment) :
+                                DRCViolation(type,
+                                             {},
+                                             shapes,
+                                             nullptr,
+                                             comment) {}
```

Comment:
clang-format

Reply:
I took a stab at this one but I wasn't sure what the clang format required for this. If I got it wrong, is there an example I can look at?

Reply:
You don't have to guess - just run clang-format -i <file> and it will be reformatted automatically.  

If you are an emacs user, I use the following:

(defun clang-format-buffer ()
  "Replace the current buffer with the result of clang-format'ing it"
  (interactive)
  (let (p start)
    (setq p (point))
    (setq start (window-start (selected-window)))
    (shell-command-on-region
     (point-min)
     (point-max)
     "/usr/bin/clang-format"
     nil ; output-buffer = current buffer
     't) ; replace
    (set-window-start (selected-window) start)
    (goto-char p)
    )
)
(global-set-key "\C-x\C-z" (quote clang-format-buffer))